File: world_ir.py
"""world_ir.py — the Python World IR (world-v1) parser/validator.

Horizon Phase 0 Build 1. Reads `grammar/protocols/world-v1.json` and validates
a world envelope (a dict of records, in Python's native value form) against
it: version first, then each of the seven §8 facets in a fixed order, then a
pass over any envelope key the protocol does not name.

Refusal follows the same voice as `interp.py`'s `records_from_json` — an
unrecognized protocol version refuses the whole envelope before any record is
checked, and every raise names a tag, a detail, and a fix. There is no
coercion at the type boundary: a field whose value fails its declared type
refuses, exactly as `+` does not coerce and `require_text` does not either.

This module adds no builtin, no effect, and no value kind — it is a validator
over the existing value model, not a change to it.
"""
import json
import os

_REPO_ROOT = os.path.dirname(os.path.abspath(__file__))
_PROTOCOL_PATH = os.path.join(_REPO_ROOT, "grammar", "protocols", "world-v1.json")

with open(_PROTOCOL_PATH, encoding="utf-8") as _f:
    PROTOCOL = json.load(_f)

SUPPORTED_VERSION = PROTOCOL["version"]
RECORDS = PROTOCOL["records"]

# Fixed enumeration order — spec §8.1-8.7, and the order every implementation
# (Python, JS, self-hosted Planes) walks the seven facets in. The conformance
# gate's byte-identical comparison depends on this order being the same
# everywhere, so it is a plain tuple, not derived from dict iteration order.
FACET_ORDER = (
    "identity", "situation", "relation", "behavior",
    "expression", "affordance", "lineage",
)
FIELD_ORDER = {facet: tuple(f["name"] for f in RECORDS[facet]["fields"]) for facet in FACET_ORDER}


class WorldIRError(Exception):
    """A refusal naming the world-v1 rule it broke — `tag`, `detail`, `fix`,
    the same three-part shape `interp.py`'s `PlanesError` carries, kept as a
    standalone class so this module has no import dependency on the language
    implementation it validates data for."""

    def __init__(self, tag, detail, fix):
        self.tag = tag
        self.detail = detail
        self.fix = fix
        super().__init__(f"{tag}: {detail}")


def _type_ok(value, type_name):
    """No coercion: a value either already matches `type_name` or it does
    not. `bool` is checked before the numeric types because Python's `bool`
    is a subclass of `int` — an unguarded numeric check would silently
    accept `true`/`false` as `integer`."""
    if type_name == "boolean":
        return isinstance(value, bool)
    if isinstance(value, bool):
        return False
    if type_name in ("identifier", "semantic-id", "source-map-path"):
        return isinstance(value, str) and len(value) > 0
    if type_name == "text":
        return isinstance(value, str)
    if type_name == "integer":
        return isinstance(value, (int, float)) and float(value).is_integer()
    if type_name == "number":
        return isinstance(value, (int, float))
    if type_name == "normalized-number":
        return isinstance(value, (int, float)) and 0 <= value <= 1
    raise WorldIRError(
        "unknown-field-type",
        f"world-v1.json declares field type {type_name!r}, which this parser does not know",
        "add a case for the new type to _type_ok in world_ir.py, in the same commit "
        "that adds it to world-v1.json")


def _validate_record(facet, record_value):
    """Returns (normalized_fields, None) or (None, reason).

    Presence is checked all at once, not field by field: a record either
    carries every field the facet declares or it does not, matching the
    self-hosted Planes implementation's `when r is { f1, f2, ... }:` —
    a single shape match, not a sequence of per-field presence tests (`when`
    cannot test a field whose name is itself a runtime value). Once presence
    holds, type failures are reported by field, in the facet's declared
    order, so every implementation names the same field first."""
    if not isinstance(record_value, dict):
        return None, "the record is not a record of fields"
    field_specs = RECORDS[facet]["fields"]
    if not all(f["name"] in record_value for f in field_specs):
        return None, "the record is missing one or more required fields"
    normalized = {}
    for field in field_specs:
        name, type_name = field["name"], field["type"]
        value = record_value[name]
        if not _type_ok(value, type_name):
            return None, f"field '{name}' fails its declared type '{type_name}'"
        normalized[name] = value
    return normalized, None


def parse_world_envelope(envelope):
    """Validate `envelope` against world-v1.

    Returns `(normalized, warnings)` on success — `normalized` is a dict
    keyed by `"version"` plus every facet present in the envelope;
    `warnings` is a sorted list of `"unknown-optional-record:<name>"`
    strings, one per envelope key the protocol does not name.

    Raises `WorldIRError` on refusal: an unsupported protocol version
    refuses the whole envelope before any record is checked; a missing or
    malformed critical record, or a malformed known-optional record, stops
    the parse. An unknown record never refuses — it warns and is dropped.
    """
    if not isinstance(envelope, dict):
        raise WorldIRError(
            "malformed-world-envelope",
            "the world envelope is not a record of fields",
            "provide the envelope as a record whose fields are 'version' plus "
            "the world-v1 facet names")

    version = envelope.get("version")
    # `isinstance(version, bool)` is checked before the equality, not folded
    # into it: Python's `bool` is a subtype of `int`, so `True == 1` is true
    # and an unguarded `!=` would silently accept `true` as version 1 — the
    # same coercion `_type_ok`'s boolean-before-numeric ordering guards
    # against for facet fields.
    if isinstance(version, bool) or version != SUPPORTED_VERSION:
        raise WorldIRError(
            "unsupported-world-protocol-version",
            f"world envelope declares protocol version {json.dumps(version)}, "
            f"which is not {SUPPORTED_VERSION}",
            "regenerate the envelope with a world-v1 protocol matching this parser's version — "
            "if the envelope is newer than what this parser reads, upgrade the parser instead "
            "of regenerating the envelope")

    normalized = {"version": SUPPORTED_VERSION}
    for facet in FACET_ORDER:
        spec = RECORDS[facet]
        if facet not in envelope:
            if spec["critical"]:
                raise WorldIRError(
                    "missing-critical-record",
                    f"critical record '{facet}' is missing from the world envelope",
                    f"add the '{facet}' record — {facet} is a critical facet and the "
                    "envelope cannot be applied without it")
            continue
        fields, reason = _validate_record(facet, envelope[facet])
        if reason is not None:
            tag = "malformed-critical-record" if spec["critical"] else "malformed-optional-record"
            raise WorldIRError(
                tag,
                f"record '{facet}' is malformed: {reason}",
                f"correct '{facet}' so every field world-v1.json declares for it is present "
                "and matches its declared type")
        normalized[facet] = fields

    known_keys = set(FACET_ORDER) | {"version"}
    warnings = sorted(
        f"unknown-optional-record:{key}" for key in envelope if key not in known_keys)

    return normalized, warnings


def _format_value(value):
    if isinstance(value, bool):
        return "true" if value else "false"
    if isinstance(value, float) and value.is_integer():
        return str(int(value))
    return str(value)


def canonical_outcome_string(envelope):
    """The cross-implementation comparison form `test_world_ir_conformance.py`
    checks byte-for-byte. Deliberately not JSON — a hand-built, fixed-order
    text form that Python, JS, and self-hosted Planes can each produce
    without needing a shared serializer, the same way the three interpreters'
    canonical-form agreement (README's 349-shape sweep) is text, not a
    structure compared by a fourth piece of code."""
    try:
        normalized, warnings = parse_world_envelope(envelope)
    except WorldIRError as e:
        return "\n".join([
            "world-ir-outcome: refuse",
            f"tag: {e.tag}",
            f"detail: {e.detail}",
            f"fix: {e.fix}",
        ])
    lines = [
        "world-ir-outcome: accept",
        f"version: {normalized['version']}",
        f"warnings: {','.join(warnings)}",
    ]
    for facet in FACET_ORDER:
        if facet not in normalized:
            continue
        for name in FIELD_ORDER[facet]:
            lines.append(f"{facet}.{name}: {_format_value(normalized[facet][name])}")
    return "\n".join(lines)


if __name__ == "__main__":
    import sys
    envelope_in = json.load(sys.stdin)
    sys.stdout.write(canonical_outcome_string(envelope_in))

File: js/world_ir.mjs
// js/world_ir.mjs — the JavaScript World IR (world-v1) parser/validator.
//
// Horizon Phase 0 Build 1. Mirrors world_ir.py field for field, refusal tag
// for refusal tag — the pair are read side by side in review, and
// test_world_ir_conformance.py is what holds them to actual byte-identical
// agreement rather than asserted agreement. Reads
// grammar/protocols/world-v1.json directly, the same as the Python side;
// grammar/world_ir.planes cannot (a .planes module has no way to read JSON),
// so its rules are hand-written there and checked against this file and
// world_ir.py by the same conformance gate.
//
// No coercion at the type boundary: a field whose value fails its declared
// type refuses, never converts.
//
// BROWSER-SAFE SINCE HORIZON PHASE 1 (renderer pipeline). This file used to
// read world-v1.json via node:fs at module load, which made it Node-only —
// invisible until the first build that needed a real browser (main thread or
// worker) to import it. The protocol is now a JSON module import, the same
// mechanism browser_main.mjs already uses for vocabulary.json/amber.json/
// core.json; PROTOCOL/SUPPORTED_VERSION/RECORDS hold the identical values
// this always produced under Node. Only the CLI mode at the bottom of this
// file (`node js/world_ir.mjs < envelope.json`, test_world_ir_conformance.py's
// shell-out target) still needs node:path/node:url, imported dynamically
// there so the static import list here stays browser-clean.

import PROTOCOL from "../grammar/protocols/world-v1.json" with { type: "json" };

export { PROTOCOL };
export const SUPPORTED_VERSION = PROTOCOL.version;
export const RECORDS = PROTOCOL.records;

// Fixed enumeration order — must match world_ir.py's FACET_ORDER exactly;
// the conformance gate's byte-identical comparison depends on it.
export const FACET_ORDER = Object.freeze([
  "identity", "situation", "relation", "behavior",
  "expression", "affordance", "lineage",
]);
export const FIELD_ORDER = Object.freeze(
  Object.fromEntries(
    FACET_ORDER.map((facet) => [facet, RECORDS[facet].fields.map((f) => f.name)]),
  ),
);

// A refusal naming the world-v1 rule it broke — the same three-part
// tag/detail/fix shape world_ir.py's WorldIRError carries.
export class WorldIRError extends Error {
  constructor(tag, detail, fix) {
    super(`${tag}: ${detail}`);
    this.tag = tag;
    this.detail = detail;
    this.fix = fix;
  }
}

function isPlainObject(v) {
  return typeof v === "object" && v !== null && !Array.isArray(v);
}

// No coercion: a value either already matches `typeName` or it does not.
// `typeof true === "boolean"` is checked first for the same reason
// world_ir.py checks `bool` before its numeric types — nothing here treats
// a boolean as a number.
function typeOk(value, typeName) {
  if (typeName === "boolean") return typeof value === "boolean";
  if (typeof value === "boolean") return false;
  if (typeName === "identifier" || typeName === "semantic-id" || typeName === "source-map-path") {
    return typeof value === "string" && value.length > 0;
  }
  if (typeName === "text") return typeof value === "string";
  if (typeName === "integer") return typeof value === "number" && Number.isFinite(value) && Number.isInteger(value);
  if (typeName === "number") return typeof value === "number" && Number.isFinite(value);
  if (typeName === "normalized-number") {
    return typeof value === "number" && Number.isFinite(value) && value >= 0 && value <= 1;
  }
  throw new WorldIRError(
    "unknown-field-type",
    `world-v1.json declares field type '${typeName}', which this parser does not know`,
    "add a case for the new type to typeOk in js/world_ir.mjs, in the same commit that adds it to world-v1.json",
  );
}

// Returns { normalized, reason }. Presence is checked all at once (matching
// grammar/world_ir.planes's `when r is { f1, f2, ... }:` — a single shape
// match, not a per-field presence test, since `when` cannot test a field
// whose name is itself a runtime value). Once presence holds, `reason` names
// the first wrong-typed field, in the facet's declared order.
function validateRecord(facet, recordValue) {
  if (!isPlainObject(recordValue)) {
    return { normalized: null, reason: "the record is not a record of fields" };
  }
  const fieldSpecs = RECORDS[facet].fields;
  const allPresent = fieldSpecs.every((f) => Object.prototype.hasOwnProperty.call(recordValue, f.name));
  if (!allPresent) {
    return { normalized: null, reason: "the record is missing one or more required fields" };
  }
  const normalized = {};
  for (const field of fieldSpecs) {
    const { name, type: typeName } = field;
    const value = recordValue[name];
    if (!typeOk(value, typeName)) {
      return { normalized: null, reason: `field '${name}' fails its declared type '${typeName}'` };
    }
    normalized[name] = value;
  }
  return { normalized, reason: null };
}

// Validate `envelope` against world-v1. Returns { normalized, warnings } on
// success; throws WorldIRError on refusal. See world_ir.py's
// parse_world_envelope for the full contract — the two are kept identical.
export function parseWorldEnvelope(envelope) {
  if (!isPlainObject(envelope)) {
    throw new WorldIRError(
      "malformed-world-envelope",
      "the world envelope is not a record of fields",
      "provide the envelope as a record whose fields are 'version' plus the world-v1 facet names",
    );
  }

  const version = Object.prototype.hasOwnProperty.call(envelope, "version") ? envelope.version : undefined;
  if (version !== SUPPORTED_VERSION) {
    throw new WorldIRError(
      "unsupported-world-protocol-version",
      `world envelope declares protocol version ${JSON.stringify(version ?? null)}, which is not ${SUPPORTED_VERSION}`,
      "regenerate the envelope with a world-v1 protocol matching this parser's version — "
        + "if the envelope is newer than what this parser reads, upgrade the parser instead "
        + "of regenerating the envelope",
    );
  }

  const normalized = { version: SUPPORTED_VERSION };
  for (const facet of FACET_ORDER) {
    const spec = RECORDS[facet];
    if (!Object.prototype.hasOwnProperty.call(envelope, facet)) {
      if (spec.critical) {
        throw new WorldIRError(
          "missing-critical-record",
          `critical record '${facet}' is missing from the world envelope`,
          `add the '${facet}' record — ${facet} is a critical facet and the `
            + "envelope cannot be applied without it",
        );
      }
      continue;
    }
    const { normalized: fields, reason } = validateRecord(facet, envelope[facet]);
    if (reason !== null) {
      const tag = spec.critical ? "malformed-critical-record" : "malformed-optional-record";
      throw new WorldIRError(
        tag,
        `record '${facet}' is malformed: ${reason}`,
        `correct '${facet}' so every field world-v1.json declares for it is present `
          + "and matches its declared type",
      );
    }
    normalized[facet] = fields;
  }

  const knownKeys = new Set([...FACET_ORDER, "version"]);
  const warnings = Object.keys(envelope)
    .filter((key) => !knownKeys.has(key))
    .map((key) => `unknown-optional-record:${key}`)
    .sort();

  return { normalized, warnings };
}

function formatValue(value) {
  if (typeof value === "boolean") return value ? "true" : "false";
  return String(value);
}

// The cross-implementation comparison form test_world_ir_conformance.py
// checks byte-for-byte against world_ir.py's canonical_outcome_string and
// grammar/world_ir.planes's own rendering.
export function canonicalOutcomeString(envelope) {
  let normalized;
  let warnings;
  try {
    ({ normalized, warnings } = parseWorldEnvelope(envelope));
  } catch (e) {
    if (!(e instanceof WorldIRError)) throw e;
    return [
      "world-ir-outcome: refuse",
      `tag: ${e.tag}`,
      `detail: ${e.detail}`,
      `fix: ${e.fix}`,
    ].join("\n");
  }
  const lines = [
    "world-ir-outcome: accept",
    `version: ${normalized.version}`,
    `warnings: ${warnings.join(",")}`,
  ];
  for (const facet of FACET_ORDER) {
    if (!(facet in normalized)) continue;
    for (const name of FIELD_ORDER[facet]) {
      lines.push(`${facet}.${name}: ${formatValue(normalized[facet][name])}`);
    }
  }
  return lines.join("\n");
}

// CLI mode — `node js/world_ir.mjs < envelope.json` — reads a JSON envelope
// from stdin and prints its canonical outcome string. Exists so
// test_world_ir_conformance.py (Python) can shell out to this
// implementation exactly as the existing test_js_*.py suites shell out to
// js/cli.mjs.
//
// node:path/node:url are imported dynamically, only inside this Node-only
// branch — a static top-level import of either would fail module resolution
// in a browser regardless of whether this branch ever runs (see the module
// docstring). `typeof process !== "undefined"` is false in every browser, so
// the dynamic imports are never even attempted there.
if (typeof process !== "undefined" && process.argv[1]) {
  const [{ default: path }, { fileURLToPath }] = await Promise.all([
    import("node:path"),
    import("node:url"),
  ]);
  if (path.resolve(process.argv[1]) === fileURLToPath(import.meta.url)) {
    const chunks = [];
    process.stdin.on("data", (c) => chunks.push(c));
    process.stdin.on("end", () => {
      const envelope = JSON.parse(Buffer.concat(chunks).toString("utf-8"));
      process.stdout.write(canonicalOutcomeString(envelope));
    });
  }
}
